modules/lang-groovy/src/main/java/org/elasticsearch/script/groovy/GroovyScriptEngineService.java

```diff
@@ -91,6 +92,11 @@
      */
     private final ClassLoader loader;
 
+    /**
+     * Ensures that the deprecation log entry for Groovy is only written on the first Groovy script compiled.
+     */
+    private AtomicBoolean isDeprecationLogged = new AtomicBoolean(false);
```

Comment:
@jdconrad You should be able to accomplish that by deprecating lang usage that is not painless in the various script using code (update, function_score, script search).

If done, I'd recommend keeping the Python/JavaScript deprecation on construction warnings. Those will appear in the log at least in the event that they're used infrequently. I'd also add that to Groovy's engine constructor to get the same behavior.